weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,344 @@
+        of the recovery code. (Default: Boolean True)
+
+    Returns: Boolean True if 2FA verification is successful, Boolean False otherwise.
+    """
+    # Strip spaces from the tfa_response; they are not valid character, and some authenticator
+    #   implementations display the code as "123 456"
+    tfa_response = tfa_response.replace(' ', '')
+    if len(tfa_response) == LENGTH_TOTP_CODE:
+        tfa_secret = d.engine.scalar("""
+            SELECT twofa_secret
+            FROM login
+            WHERE userid = %(userid)s
+        """, userid=userid)
+        # Validate supplied 2FA response versus calculated current TOTP value.
+        totp = pyotp.TOTP(tfa_secret)
+        # Return the response of the TOTP verification; True/False
+        return totp.verify(tfa_response)
+    elif len(tfa_response) == LENGTH_RECOVERY_CODE:
+        # Check if `tfa_response` is valid recovery code; consume according to `consume_recovery_code`,
+        #  and return True if valid, False otherwise
+        return is_recovery_code_valid(userid, tfa_response, consume_recovery_code)
+    # Other lengths are invalid; return False
+    else:
+        return False
+
+
+def get_number_of_recovery_codes(userid):
+    """
+    Get and return the number of remaining recovery codes for `userid`.
+
+    Parameters:
+        userid: The userid for which to check the count of recovery codes.
+
+    Returns:
+        An integer representing the number of remaining recovery codes.
+    """
+    return d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = %(userid)s
+    """, userid=userid)
+
+
+def generate_recovery_codes():
+    """
+    Generate a set of valid recovery codes.
+
+    Character set is defined by `libweasyl.security`, (ASCII+Numbers), limited to uppercase via
+    .upper() for readability.
+
+    Parameters: None
+
+    Returns: A set of length `_TFA_RECOVERY_CODES` where each code is `LENGTH_RECOVERY_CODE`
+    characters in length.
+    """
+    return {security.generate_key(LENGTH_RECOVERY_CODE).upper() for i in range(_TFA_RECOVERY_CODES)}
+
+
+def store_recovery_codes(userid, recovery_codes):
+    """
+    Store generated recovery codes into the recovery code table, checking for validity.
+
+    Parameters:
+        userid: The userid to save the recovery codes for.
+        recovery_codes: Comma separated unicode string of recovery codes.
+
+    Returns: Boolean True if the codes were successfully saved to the database, otherwise
+    Boolean False
+    """
+    # Force the incoming string to uppercase, then split into a list
+    codes = recovery_codes.upper().split(',')
+    # The list must exist and be equal to the current codes to generate
+    if len(codes) != _TFA_RECOVERY_CODES:
+        return False
+    # Make sure all codes are `LENGTH_RECOVERY_CODE` characters long, as expected
+    for code in codes:
+        if len(code) != LENGTH_RECOVERY_CODE:
+            return False
+
```

Comment:
I was thinking about these recovery codes being stored plaintext, and this answered the question I had here... Draft NIST SP 800-63B, "5.1.2.2. Look-up Secret Verifiers":

_"Verifiers SHALL store look-up secrets_[Note 1] _in a form that is resistant to offline attacks. Secrets SHALL be hashed with a salt value using an approved hash function as described in [SP 800-132]. The salt value SHALL be a 128-bit or longer random value generated by an approved random bit generator that is stored along with the hash result. A keyed hash function (e.g., HMAC [FIPS198-1]), with the key stored separately from the hashed authenticators (e.g., in a hardware security module) SHOULD be used to further resist dictionary attacks against the stored hashed authenticators."_
**Note 1)** _A common application of look-up secrets is the use of "recovery keys" stored by the subscriber for use in the event another authenticator is lost or malfunctions._

TL,DR: Unless you think otherwise, @charmander, note to self: These probably should be run through a bcrypt cycle. At minimum, this would require the following:

1. In ``store_recovery_codes()``: After pre-checking to ensure the codes are valid, run the codes through bcrypt to hash them before storing them. Optionally, reduce the work-factor slightly, but I don't know how much of a trade-off we would want or need here, keeping in mind the changes needed in ``is_recovery_code_valid()``.
2. In ``is_recovery_code_valid()``: Instead of straight up deleting the code if it is valid and returning it, we'd need to retrieve all current codes, then attempt to verify the provided recovery code in turn against all bcrypt hashes retrieved from the table. Upon finding a match, delete that bcrypt value/userid pair. [This might be a good case for slightly reducing the work-factor, since I am not quite certain how long it takes to run through ten bcrypt cycles back-to-back (worst case).]
3. In the ``twofa_recovery_codes`` table (both the model and the Alembic revision): The length of ``recovery_code`` would need to be extended to match the length of the bcrypt hash (I'd likely just steal the length from ``authbcrypt.hashsum``, which is 100).

The following is recommended as well:
1. In ``weasyl/controllers/two_factor_auth.py:tfa_init_verify_post_()``: This function may need to be looked at to check about the timing of storing the recovery codes versus the ``tfa.activate()`` call. Alternatively, in this module in ``activate()``, change the ``if totp.verify(tfa_response)`` call (L117 as of this comment), to ``if totp.verify(tfa_response, valid_window=1)``, or at most 2 to give a buffer window for the bcrypt iterations over the recovery codes to complete.

I am going to ignore the bit where it says only PBKDF should be used as the hash derivation function (SP 800-132), as we are not bound to use NIST guidelines; I would, however, say that salt-hashing these is likely a Very Good Idea™.

As a side benefit, if we opt to change the recovery code length in the future, no further logic would be required to handle the updated length. Rather, we'd just need to alter the length of newly generated codes, and since input is hashed, we don't really care about input length.

Reference:
https://pages.nist.gov/800-63-3/sp800-63b.html#sec5